runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu

```diff
@@ -77,6 +83,252 @@ __global__ void setFirstNElements(T *sv, const T *__restrict__ sv2, int64_t N) {
+        /*maskLen*/ 0, &extraWorkspaceSizeInBytes));
+    // allocate external workspace if necessary
+    void *extraWorkspace = nullptr;
+    if (extraWorkspaceSizeInBytes > 0)
+      HANDLE_CUDA_ERROR(cudaMalloc(&extraWorkspace, extraWorkspaceSizeInBytes));
+
+    // set external workspace
+    HANDLE_ERROR(custatevecAccessorSetExtraWorkspace(
+        handle, accessor, extraWorkspace, extraWorkspaceSizeInBytes));
+
+    HANDLE_ERROR(
+        custatevecAccessorGet(handle, accessor, &value[0], start, end));
+    // destroy descriptor
+    HANDLE_ERROR(custatevecAccessorDestroy(accessor));
+    // free extra workspace if allocated
+    if (extraWorkspaceSizeInBytes > 0)
+      HANDLE_CUDA_ERROR(cudaFree(extraWorkspace));
+  }
+
+  bool isDevicePointer(void *ptr) const {
+    cudaPointerAttributes attributes;
+    cudaPointerGetAttributes(&attributes, ptr);
+    return attributes.type > 1;
+  }
+
+  std::size_t deviceFromPointer(void *ptr) const {
+    cudaPointerAttributes attributes;
+    cudaPointerGetAttributes(&attributes, ptr);
+    return attributes.device;
+  }
+
+public:
+  CusvState(std::size_t s, void *ptr) : size(s), devicePtr(ptr) {
+    custatevecCreate(&handle);
+  }
+
+  CusvState(const std::vector<std::size_t> &shape,
+            const std::vector<std::complex<double>> &data) {
+    if (shape.size() != 1)
+      throw std::runtime_error(
+          "CusvState must be created from data with 1D shape.");
+
+    size = shape[0];
+    cudaMalloc((void **)&devicePtr, size * sizeof(std::complex<ScalarType>));
+    cudaMemcpy(devicePtr, &data[0],
+               data.size() * sizeof(std::complex<ScalarType>),
+               cudaMemcpyHostToDevice);
+  }
+
+  std::size_t getNumQubits() const override { return std::log2(size); }
+  std::vector<std::size_t> getDataShape() const override { return {size}; }
+
+  double overlap(const cudaq::SimulationState &other) override {
+    if (getDataShape() != other.getDataShape())
+      throw std::runtime_error("[custatevec-state] overlap error - other state "
+                               "dimension not equal to this state dimension.");
+
+    // Check if input vector is on the right GPU device,
+    // if so, compute on device
+    thrust::device_ptr<ThrustComplex<ScalarType>> thrustDevPtrABegin(
+        reinterpret_cast<ThrustComplex<ScalarType> *>(devicePtr));
+    thrust::device_ptr<ThrustComplex<ScalarType>> thrustDevPtrAEnd(
+        reinterpret_cast<ThrustComplex<ScalarType> *>(devicePtr) + size);
+
+    // Make sure other is on GPU device already
+    if (isDevicePointer(other.ptr())) {
+      if (deviceFromPointer(devicePtr) != deviceFromPointer(other.ptr()))
+        throw std::runtime_error(
+            "overlap requested for device pointers on separate GPU devices.");
+      // other is a device pointer
+      thrust::device_ptr<ThrustComplex<ScalarType>> thrustDevPtrBBegin(
+          reinterpret_cast<ThrustComplex<ScalarType> *>(other.ptr()));
```

Comment:
Q: would `thrust` handle different data type implicitly?
e.g.,  `other.getPrecision() != this->getPrecision()`

Reply:
Updated to be a bit more explicit about precision, added more error checking